### DISCUSSION TITLE
New addon to import custom blocks from a already existing project

### DISCUSSION THREAD

**Discussion by ghost** [ScratchAddons/ScratchAddons#4793] [Ideas]
Now, I know this may sound a little if-y, but hear out this idea. I've been using text languages for a long while, and anyone who uses Python probably knows what the `import` keyword does. I've long been awaiting a time for when Scratch adopts using libraries that use custom blocks (functions) pulled from another project and imports them into my new project. I would like to know if I can begin developing a new addon that makes use of this. 

If using the same method as the debugger for importing blocks is not an appreciated way of getting this to work, how about editing the JSON of a project to "inject" custom blocks into it.

If this is not a plausible addon idea, please let me know, but if there are ways to do this that I have not though of, please, let me know if I can start developing.

**Comment by WorldLanguages** [comment]
This in theory would be great, but it's still complex and it needs to be intuitive to use. But I don't think there would be a problem by adding a feature that automatically imports custom blocks from other projects. The backpack can already achieve that, but it's not automated so it has to be done manually, which is slower and more annoying.

**Reply by CST1229** [reply]
Design idea:
1. A new Import Custom Blocks option is added in the File menu. Clicking on this opens a modal with 3 screens.
2. On the first screen, you have to input a project ID or URL, or pick from projects that you recently imported blocks from.
3. On the second screen, you select all the custom blocks to import. The screen is divided into tabs for each sprite (and the stage) in the project you selected in the previous screen, with a list of checkboxes for each custom block in each tab.  
If a block depends on another block, that block is automatically selected, too.  
Additionally, all variables that are required by the blocks will be imported, but those are not shown yet.
5. On the third screen, you select which sprite in the project to import the blocks into, and a confirmation showing all the custom blocks and variables which will be imported. Confirming the import closes the modal and imports the blocks into the sprite.

**Reply by ghost** [reply]
Sounds great. It will take a while, but I've already started poking around SB3 files to see how it would work, and the way everything is laid out makes it pretty simple. I don't quite know how to fetch the SB3 files from Scratch's API/CD2 networks, but I was thinking of using Forkphorus's [project downloader](https://forkphorus.github.io/sb-downloader/) source code.

**Reply by WorldLanguages** [reply]
Manually editing the project JSON is probably not the right way to try to tackle this as an addon. I don't think any editor addons directly modify the project JSON. I don't think you can apply the updates without a reload either.

**Reply by ghost** [reply]
That may be so. Would there be any real security risk or downside to pulling the JSON, editing it/injecting the custom blocks, then using a Scratch editor JS function to load the new JSON into the editor
